runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/internal/registry/FileWatcherTask.java

```diff
@@ -0,0 +1,140 @@
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static org.agrona.LangUtil.rethrowUnchecked;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+
+
+public class FileWatcherTask extends WatcherTask
+{
+    private final Map<Path, byte[]> configHashes;
+    private final Map<Path, URL> configURLs;
+    private MessageDigest md5;
+    private WatchService watchService;
+
+    public FileWatcherTask(
+        BiConsumer<URL, String> configChangeListener)
+    {
+        super(configChangeListener);
+        this.configHashes = new HashMap<>();
+        this.configURLs = new HashMap<>();
+        try
+        {
+            this.md5 = MessageDigest.getInstance("MD5");
+            this.watchService = FileSystems.getDefault().newWatchService();
+        }
+        catch (NoSuchAlgorithmException | IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+
+    }
+
+    @Override
+    public boolean run()
+    {
+        try
+        {
+            while (true)
+            {
+                try
+                {
+                    final WatchKey key = watchService.take();
+
+                    for (WatchEvent<?> event : key.pollEvents())
+                    {
+                        final Path changed = ((Path) key.watchable()).resolve((Path) event.context());
+                        if (configHashes.containsKey(changed))
+                        {
+                            String newConfigText = readConfigText(configURLs.get(changed));
+                            byte[] oldConfigHash = configHashes.get(changed);
+                            byte[] newConfigHash = computeHash(newConfigText);
+                            if (!Arrays.equals(oldConfigHash, newConfigHash))
+                            {
+                                configHashes.put(changed, newConfigHash);
+                                configChangeListener.accept(configURLs.get(changed), newConfigText);
+                            }
+                        }
+                    }
+                    key.reset();
+                }
+                catch (InterruptedException ex)
+                {
+                    watchService.close();
+                    break;
+                }
+            }
+        }
+        catch (IOException ex)
+        {
+            rethrowUnchecked(ex);
+        }
+        return true;
+    }
+
+    @Override
+    protected void doInitialConfiguration(
```

Comment:
There's no such thing as initial configuration now, right?
Let's rename this, and also determine if it is still needed on the superclass.

Reply:
That's right, we can actually remove it. But, we'll need a new way for Zilla to wait for the initial configuration. I have ideas. Let's discuss this.